File: app/src/main/java/com/sanketvagadiya/ui/Projects/ProjectsFragment.java
package com.sanketvagadiya.ui.Projects;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModelProviders;

import android.content.Context;
import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Adapter;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.Response;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.JsonObjectRequest;
import com.android.volley.toolbox.Volley;
import com.bumptech.glide.Glide;
import com.bumptech.glide.request.RequestOptions;
import com.sanketvagadiya.Constants.Constants;
import com.sanketvagadiya.Model.ProjectModel;
import com.sanketvagadiya.R;
import com.sanketvagadiya.ui.BlankViewModel;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;

public class ProjectsFragment extends Fragment {

    private ProjectsViewModel mViewModel;
    private Context mContext;
    private List<ProjectModel> projectList = new ArrayList<>();
    private RequestQueue mQueue;
    private JSONObject jsonObject = new JSONObject();
    private ListAdapter adapter;
    private ListView lvProjectList;

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        mContext = getActivity();

        View root = inflater.inflate(R.layout.projects_fragment, container, false);

        lvProjectList = root.findViewById(R.id.lvProjectList);
        return root;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        mQueue = Volley.newRequestQueue(getContext());
        sendProjectListJson();

    }


    public void sendProjectListJson() {

        JsonObjectRequest request = new JsonObjectRequest(Request.Method.GET, Constants.PROJECT_URL, null, new Response.Listener<JSONObject>() {
            @Override
            public void onResponse(JSONObject response) {
                jsonObject = response;
            }
        }, new Response.ErrorListener() {
            @Override
            public void onErrorResponse(VolleyError error) {
                Log.e("err", "err" + error);
            }
        });


        mQueue.addRequestFinishedListener(
                new RequestQueue.RequestFinishedListener<Object>() {
                    @Override
                    public void onRequestFinished(Request<Object> request) {
                        try {
                            JSONArray jsonArray = jsonObject.getJSONArray(Constants.PROJECT_JSONARRAY);

                            for (int i = 0; i < jsonArray.length(); i++) {
                                JSONObject jsonObject = jsonArray.getJSONObject(i);
                                String projectName = jsonObject.getString(Constants.PROJECT_JSON_PROJECTNAME);
                                String imageUrl = jsonObject.getString(Constants.PROJECT_JSON_IMAGEURL);
                                String projectUrl = jsonObject.getString(Constants.PROJECT_JSON_PROJECTURL);
                                projectList.add(new ProjectModel(projectName, imageUrl, projectUrl));
                            }
                            if (projectList.size() > 0) {
                                adapter = new ListAdapter();
                                lvProjectList.setAdapter(adapter);
                            }

                        } catch (JSONException e) {
                            e.printStackTrace();
                        }

                    }
                }
        );
        mQueue.add(request);
    }


    public class ListAdapter extends BaseAdapter {
        @Override
        public int getCount() {
            return projectList.size();
        }

        @Override
        public Object getItem(int i) {
            return i;
        }

        @Override
        public long getItemId(int i) {
            return i;
        }

        @Override
        public View getView(int i, View convertView, ViewGroup viewGroup) {

            final ViewHolder holder;

            LayoutInflater inflater = (LayoutInflater)
                    mContext.getSystemService(mContext.LAYOUT_INFLATER_SERVICE);
            if (convertView == null) {
                holder = new ViewHolder();

                convertView = inflater.inflate(R.layout.project_listview_adapter, viewGroup, false);

                holder.tvProjectName = (TextView) convertView.findViewById(R.id.tvProjectName);
                holder.ivProjectImage = (ImageView) convertView.findViewById(R.id.ivProjectImage);

                holder.tvProjectName.setText(projectList.get(i).getProjectName());

                Glide.with(getContext())
                        .load(projectList.get(i).getImageUrl())
                        .apply(RequestOptions.circleCropTransform())
                        .into(holder.ivProjectImage);
                convertView.setTag(holder);

            } else {
                holder = (ViewHolder) convertView.getTag();
            }

            return convertView;

        }
    }

    public class ViewHolder {
        TextView tvProjectName;
        ImageView ivProjectImage;
    }


}

File: app/src/main/java/com/sanketvagadiya/ui/Projects/ProjectsFragment.kt
package com.sanketvagadiya.ui.Projects

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import androidx.fragment.app.Fragment

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast

import com.android.volley.Request
import com.android.volley.RequestQueue
import com.android.volley.Response
import com.android.volley.VolleyError
import com.android.volley.toolbox.JsonObjectRequest
import com.android.volley.toolbox.Volley
import com.bumptech.glide.Glide
import com.bumptech.glide.request.RequestOptions
import com.sanketvagadiya.Constants.Constants
import com.sanketvagadiya.Model.ProjectModel
import com.sanketvagadiya.R

import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList

class ProjectsFragment : Fragment() {

    private val mViewModel: ProjectsViewModel? = null
    private var mContext: Context? = null
    private val projectList = ArrayList<ProjectModel>()
    private var mQueue: RequestQueue? = null
    private var jsonObject = JSONObject()
    private var adapter: ListAdapter? = null
    private var lvProjectList: ListView? = null

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        mContext = activity

        val root = inflater.inflate(R.layout.projects_fragment, container, false)

        lvProjectList = root.findViewById(R.id.lvProjectList)
        return root
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        mQueue = Volley.newRequestQueue(context!!)
        sendProjectListJson()

    }


    fun sendProjectListJson() {
        val request = JsonObjectRequest(Request.Method.GET, Constants.PROJECT_URL, null, Response.Listener { response -> jsonObject = response }, Response.ErrorListener { error -> Log.e("err", "err$error") })


        mQueue!!.addRequestFinishedListener(
                RequestQueue.RequestFinishedListener<Any> {
                    try {
                        val jsonArray = jsonObject.getJSONArray(Constants.PROJECT_JSONARRAY)

                        for (i in 0 until jsonArray.length()) {
                            val jsonObject = jsonArray.getJSONObject(i)
                            val projectName = jsonObject.getString(Constants.PROJECT_JSON_PROJECTNAME)
                            val imageUrl = jsonObject.getString(Constants.PROJECT_JSON_IMAGEURL)
                            val projectUrl = jsonObject.getString(Constants.PROJECT_JSON_PROJECTURL)
                            projectList.add(ProjectModel(projectName, imageUrl, projectUrl))
                        }
                        if (projectList.size > 0) {
                            adapter = ListAdapter()
                            lvProjectList!!.adapter = adapter
                        }

                    } catch (e: JSONException) {
                        e.printStackTrace()
                    }
                }
        )
        mQueue!!.add(request)
    }


    inner class ListAdapter : BaseAdapter() {
        override fun getCount(): Int {
            return projectList.size
        }

        override fun getItem(i: Int): Any {
            return i
        }

        override fun getItemId(i: Int): Long {
            return i.toLong()
        }

        override fun getView(i: Int, convertView: View?, viewGroup: ViewGroup): View {
            var convertView = convertView

            val holder: ViewHolder

//            val inflater = mContext!!.getSystemService(mContext!!.LAYOUT_INFLATER_SERVICE) as LayoutInflater
            val inflater = LayoutInflater.from(getActivity());


            if (convertView == null) {
                holder = ViewHolder()

                convertView = inflater.inflate(R.layout.project_listview_adapter, viewGroup, false)

                holder.tvProjectName = convertView!!.findViewById<View>(R.id.tvProjectName) as TextView
                holder.ivProjectImage = convertView.findViewById<View>(R.id.ivProjectImage) as ImageView

                holder.tvProjectName!!.text = projectList[i].projectName

                Glide.with(context!!)
                        .load(projectList[i].imageUrl)
                        .apply(RequestOptions.circleCropTransform())
                        .into(holder.ivProjectImage!!)


                convertView.setOnClickListener {
                    val uri = Uri.parse(projectList[i].projectUrl) // missing 'http://' will cause crashed
                    val intent = Intent(Intent.ACTION_VIEW, uri)
                    startActivity(intent)
                }


                convertView.tag = holder

            } else {
                holder = convertView.tag as ViewHolder
            }

            return convertView

        }
    }

    inner class ViewHolder {
        internal var tvProjectName: TextView? = null
        internal var ivProjectImage: ImageView? = null
    }


}
